File: src-tauri/src/analysis/error_hotspots/shared.rs
use std::path::PathBuf;

use serde::Deserialize;

use crate::discovery::path_decoder;
use crate::watcher;

const ERROR_PREFIX_LEN: usize = 100;

#[derive(Deserialize)]
pub(super) struct RawEntry {
    pub(super) timestamp: Option<String>,
    pub(super) message: Option<RawMessage>,
}

#[derive(Deserialize)]
pub(super) struct RawMessage {
    pub(super) role: Option<String>,
    pub(super) content: Option<serde_json::Value>,
}

pub(super) struct ToolCall {
    pub(super) tool_name: String,
}

pub(super) fn parse_timestamp_ms(ts: &str) -> Option<f64> {
    chrono::DateTime::parse_from_rfc3339(ts)
        .ok()
        .map(|dt| dt.timestamp_millis() as f64)
}

pub(super) fn tool_result_text(content: &serde_json::Value) -> String {
    match content {
        serde_json::Value::String(s) => s.clone(),
        serde_json::Value::Array(arr) => arr
            .iter()
            .filter_map(|block| {
                let kind = block.get("type")?.as_str()?;
                if kind == "text" {
                    block.get("text")?.as_str().map(|s| s.to_string())
                } else {
                    None
                }
            })
            .collect::<Vec<_>>()
            .join("\n"),
        _ => String::new(),
    }
}

pub(super) fn normalize_error_prefix(text: &str) -> String {
    let trimmed = text.trim();
    let clipped: String = trimmed.chars().take(ERROR_PREFIX_LEN).collect();
    clipped.split_whitespace().collect::<Vec<_>>().join(" ")
}

pub(super) fn resolve_project_dir(project_id: &str) -> Result<PathBuf, String> {
    let claude_dir = watcher::resolve_claude_dir().ok_or("Cannot resolve home directory")?;
    let projects_dir = path_decoder::get_projects_base_path(&claude_dir);
    let base_id = match project_id.find("::") {
        Some(idx) => &project_id[..idx],
        None => project_id,
    };
    let project_dir = projects_dir.join(base_id);
    if !project_dir.is_dir() {
        return Err(format!("Project directory not found: {base_id}"));
    }
    Ok(project_dir)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalizes_prefix() {
        assert_eq!(normalize_error_prefix("  hello   world  "), "hello world");
        let long = "a".repeat(200);
        assert_eq!(
            normalize_error_prefix(&long).chars().count(),
            ERROR_PREFIX_LEN
        );
    }
}

File: internal/insights/error_hotspots/shared.go
package error_hotspots

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"
	"unicode/utf8"

	"claude-devtools/internal/discovery"
)

// rawEntry is the minimal JSONL shape for error hotspot scanning.
type rawEntry struct {
	Timestamp *string `json:"timestamp"`
	Message   *rawMsg `json:"message"`
}

type rawMsg struct {
	Role    *string          `json:"role"`
	Content *json.RawMessage `json:"content"`
}

// toolCall tracks an in-flight tool_use awaiting its tool_result.
type toolCall struct {
	toolName string
}

// parseTimestampMs parses an RFC-3339 timestamp to ms since epoch.
func parseTimestampMs(ts string) (float64, bool) {
	t, err := time.Parse(time.RFC3339Nano, ts)
	if err != nil {
		t2, err2 := time.Parse(time.RFC3339, ts)
		if err2 != nil {
			return 0, false
		}
		return float64(t2.UnixMilli()), true
	}
	return float64(t.UnixMilli()), true
}

// toolResultText extracts text from a tool_result content field.
func toolResultText(raw json.RawMessage) string {
	var s string
	if json.Unmarshal(raw, &s) == nil {
		return s
	}
	var blocks []json.RawMessage
	if json.Unmarshal(raw, &blocks) != nil {
		return ""
	}
	var parts []string
	for _, b := range blocks {
		var block struct {
			Type string `json:"type"`
			Text string `json:"text"`
		}
		if json.Unmarshal(b, &block) == nil && block.Type == "text" {
			parts = append(parts, block.Text)
		}
	}
	return strings.Join(parts, "\n")
}

// normalizeErrorPrefix trims, clips to errorPrefixLen runes, and collapses
// whitespace. Mirrors shared.rs::normalize_error_prefix.
func normalizeErrorPrefix(text string) string {
	trimmed := strings.TrimSpace(text)
	// Clip to errorPrefixLen runes.
	runes := []rune(trimmed)
	if len(runes) > errorPrefixLen {
		// Ensure we don't split a multibyte char — use utf8 safe clip.
		clipped := make([]byte, 0, errorPrefixLen*4)
		count := 0
		for i := 0; i < len(trimmed) && count < errorPrefixLen; {
			r, size := utf8.DecodeRuneInString(trimmed[i:])
			_ = r
			clipped = append(clipped, trimmed[i:i+size]...)
			i += size
			count++
		}
		trimmed = string(clipped)
	}
	// Collapse whitespace.
	fields := strings.Fields(trimmed)
	return strings.Join(fields, " ")
}

// resolveProjectDir resolves the project directory from a project ID.
func resolveProjectDir(projectID string) (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("cannot resolve home directory")
	}
	claudeDir := home + "/.claude"
	projectsDir := discovery.GetProjectsBasePath(claudeDir)

	baseID := projectID
	if i := strings.Index(projectID, "::"); i >= 0 {
		baseID = projectID[:i]
	}
	dir := projectsDir + "/" + baseID
	info, err2 := os.Stat(dir)
	if err2 != nil || !info.IsDir() {
		return "", fmt.Errorf("project directory not found: %s", baseID)
	}
	return dir, nil
}
